cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/utils.go

```diff
@@ -143,6 +130,161 @@ func newPackageRepository(unstructuredRepo map[string]interface{}) (*v1alpha1.Pa
+	pkg.IconUrl = chart.Icon
+	pkg.ShortDescription = chart.Description
+
+	pkg.AvailablePackageRef = &corev1.AvailablePackageReference{
+		Identifier: chart.ID,
+		Plugin:     GetPluginDetail(),
+	}
+	pkg.AvailablePackageRef.Context = &corev1.Context{Namespace: chart.Repo.Namespace}
+
+	if chart.ChartVersions != nil || len(chart.ChartVersions) != 0 {
+		pkg.LatestPkgVersion = chart.ChartVersions[0].Version
+	}
+
+	return pkg, nil
+}
+
+func passesFilter(chart chart.Chart, filters *corev1.FilterOptions) bool {
+	if filters == nil {
+		return true
+	}
+	ok := true
+	if categories := filters.GetCategories(); len(categories) > 0 {
+		ok = false
+		for _, cat := range categories {
+			if cat == chart.Category {
+				ok = true
+				break
+			}
+		}
+	}
+	if ok {
+		if appVersion := filters.GetAppVersion(); len(appVersion) > 0 {
+			ok = appVersion == chart.ChartVersions[0].AppVersion
+		}
+	}
+	if ok {
+		if pkgVersion := filters.GetPkgVersion(); len(pkgVersion) > 0 {
+			ok = pkgVersion == chart.ChartVersions[0].Version
+		}
+	}
+	if ok {
+		if query := filters.GetQuery(); len(query) > 0 {
+			ok = strings.Contains(chart.Name, query)
+			// TODO (gfichtenholt) possibly also chart keywords, but it has not been clarified yet
+		}
+	}
+	return ok
+}
+
+// pageOffsetFromPageToken converts a page token to an integer offset
+// representing the page of results.
+// TODO(gfichtenholt): it'd be better if we ensure that the page_token
+// contains an offset to the item, not the page so we can
+// aggregate paginated results. Same as helm hlug-in.
+// Update this when helm plug-in does so
+func pageOffsetFromPageToken(pageToken string) (int, error) {
+	if pageToken == "" {
+		return 1, nil
+	}
+	offset, err := strconv.ParseUint(pageToken, 10, 0)
+	if err != nil {
+		return 0, err
+	}
+
+	return int(offset), nil
+}
+
+func getPaginatedSummariesWithFilters(pageSize, pageOffset int, cachedCharts map[string]interface{}, filters *corev1.FilterOptions) ([]*corev1.AvailablePackageSummary, error) {
+	// this loop is here for 3 reasons:
+	// 1) to convert from []interface{} which is what the generic cache implementation
+	// returns for cache hits to a typed array object.
+	// 2) perform any filtering of the results as needed, pending redis support for
+	// querying values stored in cache (see discussion in https://github.com/kubeapps/kubeapps/issues/3032)
+	// 3) if pagination was requested, only return up to one page size of results
+	responsePackages := make([]*corev1.AvailablePackageSummary, 0)
+	i := 0
+	startAt := -1
+	if pageSize > 0 {
+		startAt = int(pageSize) * pageOffset
+	}
+	for _, packages := range cachedCharts {
+		if packages != nil {
+			typedCharts, ok := packages.([]chart.Chart)
+			if !ok {
+				return nil, status.Errorf(
+					codes.Internal,
+					"Unexpected value fetched from cache: %v", packages)
+			} else {
+				for _, chart := range typedCharts {
+					if passesFilter(chart, filters) {
+						i++
+						if startAt < 0 || startAt < i {
```

Comment:
Isn't that a tautology? Given `i >= 0`, it can just be `if startAt < i {`?

Reply:
yeah, you're right. I first wrote the case for when no pagination was specified and then later added a clause for when it was. I will simplify